### ISSUE TITLE
Multi language / i18n Support

### ISSUE THREAD

**Issue by jrondorf** [misterunknown/ifm#64]
I am preparing the multi language support. I created a prototype with some translations to see if it works.

Please check https://github.com/jrondorf/ifm/commit/438af47ffa38a67b142fac6bbcaecaf300178cec if this looks basically good to you.

Embedding only one language at a time is not covered currently.

**Comment by misterunknown** [comment]
Thanks for the great effort! This is exactly what I thought of, and how I would have implemented it myself. Don't worry about the one-language-at-a-time problem, I'll add this feature later to the compiler. I plan to create a little website where you can create and download custom IFM builds.

I'd change only one little thing: The name of the config variable should be `language` instead of `locale`, as the term _locale_ mostly refers not only to the actual language, but also to other relevant settings like encoding etc. As the IFM forces and depends on UTF-8 `language` would be the better name. Let me know if you disagree ;)

I just created the branch [i18n](https://github.com/misterunknown/ifm/tree/i18n), which uses jrondorf/master as the upstream. But anyway, feel free to create a pull request, once you consider it done.

**Comment by jrondorf** [comment]
Created some new translations: https://github.com/jrondorf/ifm/commit/d9a2251abf88ac057fdad76dc2e15da366181968

Blocker: The login form and all the modals are not rendered by Mustache.js. Is it possible to render these also, so that the {{i18n.xxx}} placeholders can be filled?

**Comment by misterunknown** [comment]
Well, you're right about the login form, because this is delivered via the PHP part.

The modals are different. Most of them didn't need a rendering yet, so I simply passed the content to the `showModal` function. I'll change that and create a pull request.

**Comment by jrondorf** [comment]
Thanks! So the login screen cannot be rendered? Then I will remove the translation references there again.

**Comment by misterunknown** [comment]
Well, I'm not sure, I'll have to think about it a bit. I don't want to include Mustache for PHP, since it would just be used for the login form. But I guess we could replace the relevant strings manually, as the login form is not that complex.

**Comment by misterunknown** [comment]
Just for reference: jrondorf/ifm#2

**Comment by jrondorf** [comment]
Template translation finished.

Todo: JS Part (maybe after JS Rewrite?)

Translate PHP Part (error messages) also?

**Comment by misterunknown** [comment]
Very nice! I pull the changes and take a look ;)

**Comment by misterunknown** [comment]
I pulled the changes and merged it into #70. The rewrite of the javascript part is not yet done, but I made a cut now, and performed a comprehensive test. Concomitant with that I changed the internal logic a bit, and consolidated some functions.

I'd really appreciate it, if you'd find the time to do some tests in your use cases. Maybe I missed some bugs.

I suggest the next steps are:
* finalize tests on #70 and fix bugs
* merge #70 to master

After that you can update your jrondorf/ifm master branch.

**Comment by misterunknown** [comment]
Well, I decided to finalize the js-rewrite, and after some hours of work it's done now. I just pushed the latest commit  to #70. I'll complete the tests over the weekend. As this is a huge update, I think about consider it a major update, so this could be IFM3.

**Comment by jrondorf** [comment]
I could perform some more translations next week then.

**Comment by jrondorf** [comment]
Did some testing. Maybe search should be performed in lowercase / case insensitive.

Also displaying the thumbnail preview of a photo inside a folder with strange special characters is not correct: 
http://ifm.jrondorf.de/ifm.php#dir!%22§%24%25%26()%3D%3F%60%23%2B'.%3B-¹²³%C2%BC%C2%BD¬%7B%5B%5D%7D%5C%3C%3E%7C_dir%2FFrgrgtgtg
-> image.jpg preview

**Comment by misterunknown** [comment]
Thanks for the work. The preview error is caused by a weird escaping error. I'll guess I should load the tooltip via a real hidden element. This should be easy to fix.

**Comment by misterunknown** [comment]
Regarding the search function: it's not easy to make it case insensitive, because I use the `glob()` function, where I can lowercase the pattern, but not the content of the directory. A possible workaround, for now, could be something like this:
```
[Ss]tuff* 
```

**Comment by misterunknown** [comment]
The escaping error is fixed with [that](https://github.com/misterunknown/ifm/pull/70/commits/6a45ae38fa03d65d06bebfc116914e9e0d651894). I already updated the version on your server. I did some tests myself, and didn't find any bugs. I do some tests on a windows machine tomorrow and then merge it to master.

**Comment by misterunknown** [comment]
I just merged #70 after doing some tests on windows. When all currently open issues are solved, I'll release version 3.0.0.

**Comment by jrondorf** [comment]
I will perform more translations tomorrow evening. I hope I will have the JS part covered with that. 

What about the PHP part, where (error) texts are added to some messages?

**Comment by misterunknown** [comment]
I think the error messages in PHP should also get translated. If you want, I can do that myself, when I solved the remaining issues.

**Comment by misterunknown** [comment]
I translated the PHP part completely, but didn't push it yet. I have to rework the error handling, because at the moment some notification messages are set in the JS part, and some error messages are set by the PHP part. I'll change this, so that all API-related messages are generated by PHP.  I think I'll push the update early next week. With that, version 3 is almost ready.

By the way: I recently added a little [demo site](https://ifmdemo.misterunknown.de).

**Comment by jrondorf** [comment]
Translation: nice!

Demo: down ;-)

**Comment by misterunknown** [comment]
Well, yes, it was a bit early to activate the demo, as I experimented with the container. It should work now ;)

**Comment by jrondorf** [comment]
Could it be that dragging and dropping of files into the file table doesn't work any more with 3.0.0-dev?

**Comment by misterunknown** [comment]
Yes, noticed that on a windows 10 firefox. The functionality works (files get moved) but you see an Firefox error page. I don't know what's wrong, yet, but I'm working on this.

**Comment by jrondorf** [comment]
I am more thinking of dragging and dropping a file from windows explorer or Finder onto the browser/filetable.

**Comment by misterunknown** [comment]
Well, thanks for the hint. The problem was caused by interfering drag and drop functions, one for the file upload and one for moving filetable items. I just fixed that, but the last commit is a mess. It contains a fix for this problem, and some server-side translations. But the german translation file now has a broken encoding, and the english translation file is incomplete. So testing the dev-3.0.0 branch in german _should_ work, but I mainly tested the drag and drop functions, yet. I'm on vacation next week, but maybe I'll get round to clean this mess up :)

**Comment by misterunknown** [comment]
A little update: Yesterday I fixed the encoding issues and updated the english language file. I still have to fix a few bugs with the notifications, but the main work should be done.

**Comment by misterunknown** [comment]
Ok. I guess the dev branch is now consistent again. I performed some tests and everything worked as expected. When I finished testing I'll merge it to the master branch. If you find any flaws let me know ;)

**Comment by misterunknown** [comment]
I finally merged dev-3.0.0 into master. I guess this issue can be closed now. Don't hesitate to create a new issue if you find anything weird.
